services/bintray/bintray-downloads.service.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const { metric } = require('../text-formatters')
+const { downloadCount } = require('../color-formatters')
+const { BaseJsonService } = require('..')
+
+const schema = Joi.object({
+  totalDownloads: Joi.number().required(),
+}).required()
+
+const versionSchema = Joi.object({
+  name: Joi.string().required(),
+}).required()
+
+module.exports = class BintrayDownloads extends BaseJsonService {
+  static get category() {
+    return 'downloads'
+  }
+
+  static get route() {
+    return {
+      base: 'bintray/d',
+      pattern: ':subject/:repo/:packageName/:version*',
+    }
+  }
+
+  static get auth() {
+    return {
+      userKey: 'bintray_user',
+      passKey: 'bintray_apikey',
+      authorizedOrigins: ['https://bintray.com'],
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Bintray',
+        staticPreview: this.render({ downloads: 69000 }),
+        namedParams: {
+          subject: 'asciidoctor',
+          repo: 'maven',
+          packageName: 'asciidoctorj',
+        },
+      },
+      {
+        title: 'Bintray (latest)',
+        staticPreview: this.render({ version: 'latest', downloads: 69000 }),
+        namedParams: {
+          subject: 'asciidoctor',
+          repo: 'maven',
+          packageName: 'asciidoctorj',
+          version: 'latest',
+        },
+      },
+      {
+        title: 'Bintray (version)',
+        staticPreview: this.render({ version: '1.6.0', downloads: 69000 }),
+        namedParams: {
+          subject: 'asciidoctor',
+          repo: 'maven',
+          packageName: 'asciidoctorj',
+          version: '1.6.0',
+        },
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'downloads' }
+  }
+
+  static render({ version, downloads }) {
+    return {
+      label: version ? `downloads@${version}` : 'downloads',
+      message: metric(downloads),
+      color: downloadCount(downloads),
+    }
+  }
+
+  async fetch({ subject, repo, packageName, version }) {
+    return this._requestJson(
+      this.authHelper.withBasicAuth({
+        schema,
+        url: version
+          ? `https://bintray.com/api/ui/version/${subject}/${repo}/${packageName}/${version}/total_downloads`
+          : `https://bintray.com/api/ui/package/${subject}/${repo}/${packageName}/total_downloads`,
```

Comment:
Do you have any links to the documentation for these APIs? We like to include them as inline comments whenever possible to help us deal with future debug/support needs

Reply:
Unfortunately no. I've reverse engineered these from what their public portal is using to fetch this data. There's another dedicated api for this, however it requires authentication even for public packages

Reply:
Yeah that's what I was afraid of. Good job on finding these (the auth wall was one of the big blockers), but I do have some hesitation with using a completely undocumented API on a product I'm not familiar with. 

I won't say it's necessarily a blocker for me just yet, but would like hear from some of the other maintainers

Reply:
Requiring authentication isn't necessarily a blocker. If we just need to be authenticated (as anyone) to see public packages and there's no rate limiting, we could run with bintray creds in production if needed.
https://github.com/badges/shields/blob/master/doc/server-secrets.md#bintray
This would be preferable to relying on an API that could change or disappear without warning.

Reply:
So we actually used to always provide creds in the Bintray requests but were regularly hitting API throttling that caused badge issues (https://github.com/badges/shields/issues/658#issuecomment-543795546)

The official API for download information requires authentication and write permissions on the package which is problematic in and of itself, but reverting back to authenticating our bintray requests would also reintroduce the throttling issues on the existing version badges :disappointed: 

https://bintray.com/docs/api/#_get_total_downloads

Reply:
OK so https://bintray.com/api/v1/packages/asciidoctor/maven/asciidoctorj/versions/_latest can be called by any user, but https://bintray.com/api/v1/packages/asciidoctor/maven/asciidoctorj/versions/1.6.0 requires auth. Wierd, but ok..

Does https://bintray.com/api/v1/packages/asciidoctor/maven/asciidoctorj/versions/1.6.0 definitely require a token with write access, or can we pass a token with read only scope? You say we need write, but my reading of:

> Security: Authenticated user with 'publish' permission for private repositories, or package read/write entitlement.

is that package read may be enough (but I'm a bit unclear if "package read/write" is one permission or "package read" and "package write" are two different permissions). If we can do it without exposing a write token, we could set it up like CircleCI/Bitrise/CodeCov/Visual Studio App Center/etc i.e: each user supplies their own token in the route. Then we shouldn't have an issue with rate limit.

Reply:
> Does https://bintray.com/api/v1/packages/asciidoctor/maven/asciidoctorj/versions/1.6.0 definitely require a token with write access, or can we pass a token with read only scope? You say we need write, but my reading of:
>
>>    Security: Authenticated user with 'publish' permission for private repositories, or package read/write entitlement.


That's a fair question that I don't know definitively so perhaps I'm wrong. I was basing that off other API's like the one we use to (anonymously) retrieve version data, and that their docs refer to this as **read entitlement**. 
https://bintray.com/docs/api/#_get_version

> Security: Authenticated user with 'read' permission for private repositories, or version read entitlement.

Obviously read/write entitlement is something that would require authentication, but I don't know what level of scoping bintray supports with the API tokens. 

https://www.jfrog.com/confluence/display/BT/Signed+URLs%2C+Entitlements+and+Access+Keys#SignedURLs,EntitlementsandAccessKeys-EntitlementsandAccessKeys

https://www.jfrog.com/confluence/display/BT/Bintray+Security#BintraySecurity-APIKeys

Reply:
If we can't do this with a single shared auth token and we can't allow users to safely specify their own token I don't think we should rely on a non-public API. This has bitten us in the past.

Reply:
@mpetuska - i presume you're a Bintray user, can you check to see if you can create an API token that only allows read access (can't be used to write/edit packages) and check if you can call the https://bintray.com/docs/api/#_get_total_downloads API with that token?

Reply:
I'm quite positive that's possible, however it will face same throttling (rate limiting per user) issues version badge faced. Not sure you guys want to go back to that.

Reply:
> I'm quite positive that's possible, however it will face same throttling (rate limiting per user) issues version badge faced. Not sure you guys want to go back to that.

Originally we were using a single Shields-owned API token for _every_ request, which is not what we're talking about here. There's a handful of services that Shields integrates with that support authentication with tokens that can be scoped to read-only operations, and in those cases the Shields badge routes accept a query parameter so that users provide their own tokens, and Shields then uses their provided token when issuing the request for their package/etc. data. 

Only individual users would be capable of exceeding the rate limits for their own account, and even if they did so it wouldn't impact other users. 

The catch to this approach though is that the tokens have to be read-only scopable, because users wouldn't want to have to include a read-write token, nor we would accept one.

Reply:
How about making token optional and if nothing is passed fallback to the current implementation?

Reply:
Just played around with the documented api and looks like that endpoint is only available for premium packages... furthermore, bintray does not allow setting up read-only tokens, so the only workaround for that would be a new service user... Looks like the current implementations is the best we have for now. Could we get this merged and revisit if/when bintray updates their apis?

Reply:
> the only workaround for that would be a new service user.

That would be a reversion to the previous state which is not acceptable. 

> Could we get this merged and revisit if/when bintray updates their apis

The impetus for this thread is that we're unlikely to merge this with reliance an undocumented/unsupported API, so we wanted to see if using documented/supported APIs was possible, which unfortunately it seems they are not.

Reply:
By service user I meant each user would have to create their own service user and use that token. But anyways, since the undocumented API is the only viable option at the moment, looks like sadly you'll have to abandon this PR.

Reply:
> By service user I meant each user would have to create their own service user and use that token

Gotcha, thank you for the clarification! The main problem there would still be the inability to have read-only tokens though.

I've also un-resolved this thread to keep it open for a bit to allow other maintainers to weigh in on whether the undocumented API is something we could accept.

Reply:
Just to confirm, all statistics endpoints of the documented api are only available for premium users (and not open-source users).

Reply:
Do we think it is intentional that the `/_latest` endpoint works then?

Reply:
@chris48s if you're talking about latest version, it's available for open-source projects and as such it does not require any auth (but works with it as well)

Reply:
As long as we're supposed to be able to call `/_latest`, I propose we add the badge, but only the latest version variant for now.

Reply:
Sorry I feel like I missed something. What are the statistics APIs in this context?

Reply:
From my perspective latest means get latest version. Statistics api here meand get download count for either all versions total, specific version or latest version. I personally need total downloads variant most

Reply:
If you want to make a badge that relies on some arbitrary JSON, there are a couple of options. You could do it with the dynamic JSON badge. e.g:

![](https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Fbintray.com%2Fapi%2Fui%2Fpackage%2Fasciidoctor%2Fmaven%2Fasciidoctorj%2Ftotal_downloads&label=total%20downloads&query=$.totalDownloads&color=blue) - https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Fbintray.com%2Fapi%2Fui%2Fpackage%2Fasciidoctor%2Fmaven%2Fasciidoctorj%2Ftotal_downloads&label=total%20downloads&query=$.totalDownloads&color=blue

or you could do it with the endpoint badge https://shields.io/endpoint and format the number a bit more nicely.